File: PlatformCode/Wrapper.swift
//
//  Wrapper.swift
//  Brogue
//
//  Created by Raymund Vidar on 10/16/19.
//  Copyright © 2019 darthvid. All rights reserved.
//

//import Foundation
// Darth: Needed for DispatchQueue
import AppKit

struct WrapperGlobals {
    static var wrapper: Wrapper? = nil
}



class Wrapper {
    let dqInputEvents = DispatchQueue(label: "InputEvents", qos: .background)
    //let glyphMgr = GlyphMgr()
    
    var isAppActive: Bool = false
    var currEvent: NSEvent? = nil
    
    init() {
        // Darth: Register C callbacks.
        callbacks.isEventWhilePaused = wrapIsEventWhilePaused
        callbacks.isAppActive = wrapIsAppActive
        callbacks.getBrogueEvent = wrapGetBrogueEvent
        callbacks.isControlKeyDown = wrapIsControlKeyDown
        callbacks.plotChar = wrapPlotChar
    }
    
    
    func setInputEvent(_ event: NSEvent) {
        //print("\(#function)")
        
        dqInputEvents.async(execute: {
            //print(event.charactersIgnoringModifiers!)
            self.currEvent = event
        })
    }
    
    func isInputEventExist() -> Bool {
        var res: Bool = false
        dqInputEvents.sync(execute: {
            res =  (nil != self.currEvent)
        })
        return res
    }
    
    func getInputEvent() -> rogueEvent {
        var evt = NSEvent()
        var ret = rogueEvent(eventType: NUMBER_OF_EVENT_TYPES, param1: 0, param2: 0, controlKey: 0, shiftKey: 0)
        
        //print("\(#function)")
        
        // Darth: Copy the event then reset. Maybe I should just have a permanent variable and just change the event status to something non-existent? Think about that for the future.
        
        dqInputEvents.sync(execute: {
            if (nil != self.currEvent) {
                evt = self.currEvent!
                self.currEvent = nil
            }
            
        })
        
        if (NSEvent.EventType.keyDown == evt.type) {
            print("\(#function) \(evt.charactersIgnoringModifiers!)")
            
            ret.eventType = KEYSTROKE
            let s = evt.charactersIgnoringModifiers!.unicodeScalars
            ret.param1 = Int(s[s.startIndex].value)
            ret.controlKey = evt.modifierFlags.contains(.control) ? 1 : 0
            ret.shiftKey = evt.modifierFlags.contains(.shift) ? 1 : 0
        }
        
        return ret
    }
    
    func runGameLoop() {
        // Darth: Run the game loop in a separate thread (background QoS) via DispatchQueue.
        
        let group = DispatchGroup()
        let dispatchQueue = DispatchQueue(label: "BrogueGameLoop", qos: .background)
        group.enter()
        dispatchQueue.async(group: group,  execute: {
            // Darth: Run Brogue's rogueMain()
            rogueMain()
            group.leave()
        })
        

        // does not wait. But the code in notify() gets after enter() and leave() calls are balanced
        group.notify(queue: .main) {
            print("\(#function): runGame() has finished!")
            self.evtBrogueEnded()
        }
    }
    
    func evtBrogueEnded() {
        // trigger the app to close
        print("\(#function)")
        exit(0)
    }
    
    func plotChar(_ ch: PlotCharStruct) {
        //print("\(#function)", ch.inputChar)
    }
}

/*
 ******* [The callbacks called by wrapper.c. Not part of class but bridges to it.] *******
 */

func wrapGetBrogueEvent(textInput: Int8, colorsDance: Int8) -> rogueEvent {
    let evt = rogueEvent(eventType: NUMBER_OF_EVENT_TYPES, param1: 0, param2: 0, controlKey: 0, shiftKey: 0)
    
    //print("\(#function),", textInput, colorsDance)
    return (WrapperGlobals.wrapper?.getInputEvent() ?? evt)
}


func wrapIsEventWhilePaused(milliseconds: Int32) -> Int8 {
    // Returns true if the player interrupted the wait with a keystroke or mouse action; otherwise false.
    
    var res: Int8 = 0
    
    //print("\(#function),", milliseconds)
    //print("\(#function),", milliseconds, WrapperGlobals.wrapper!.isAppActive)
    if (WrapperGlobals.wrapper!.isAppActive) {
        // Darth: Use DQ to check. Return "true" if there is an event available.
        if (WrapperGlobals.wrapper!.isInputEventExist()) {
            res = Int8(truncating: NSNumber(true))
        }
        else {
            // Darth: Should we actually implement some sort of sleep here? ¯\_(ツ)_/¯
            //          How to provide that delay without killing the thread?
            //sleep(UInt32(milliseconds))
            //print("fake sleep,", milliseconds, WrapperGlobals.wrapper!.isAppActive)
            DispatchQueue.main.asyncAfter(deadline: .now() + 0.05) {
                // your code here
            }
        }
    }
    
    return res
}

func wrapIsAppActive() -> Int8 {
    //print("\(#function)", WrapperGlobals.wrapper!.isAppActive)
    return Int8(truncating: NSNumber(value:WrapperGlobals.wrapper!.isAppActive))
}

func wrapIsControlKeyDown() -> Int8 {
    // Darth: There's another solution assuming one needs to track this all throughout, but this is only needed in the main menu, so perhaps it's worth keeping the current polling solution.
    
    var res = Int8(0)
    DispatchQueue.main.async {
        //Do UI Code here.
        res = ((NSApp.currentEvent!.modifierFlags.contains(.control)) ? 1 : 0)
    }
    if (0 != res) {
        print("\(#function)", res)
    }
    
    return res
}

func wrapPlotChar(ch: PlotCharStruct) {
    //print("\(#function)", ch.inputChar)
    WrapperGlobals.wrapper?.plotChar(ch)
}

File: PlatformCode/Wrapper.c
//
//  Wrapper.m
//  Brogue
//
//  Created by Raymund Vidar on 10/11/19.
//  Copyright © 2019 darthvid. All rights reserved.
//

//#import <Foundation/Foundation.h>
#include "Rogue.h"
#include "Wrapper.h"


/*
Complete missing C functions from BrogueCode then transfer calls to Wrapper.swift so all actual handling is via Swift.
"callbacks" struct is actually initialized in Wrapper.swift, but has to be declared here for C functions to find the global.
*/

cbStruct callbacks;



boolean controlKeyIsDown() {
    // Darth: Seems only used for the main menu to change "New Game" to "New Game Custom".
    
    //printf("%s: \n", __PRETTY_FUNCTION__);
    //return false;
    return callbacks.isControlKeyDown();
}

short getHighScoresList(rogueHighScoresEntry returnList[HIGH_SCORES_COUNT]) {
    // returns the index number of the most recent score
    
    printf("%s: \n", __PRETTY_FUNCTION__);
    return 0;
}

void initializeBrogueSaveLocation() {
    printf("%s: \n", __PRETTY_FUNCTION__);
}

void initializeLaunchArguments(enum NGCommands *command, char *path, unsigned long *seed) {
    printf("%s: \n", __PRETTY_FUNCTION__);
}

boolean isApplicationActive() {
    // Darth: Seems only used for the main menu to change "New Game" to "New Game Custom".
    
    //printf("%s(%d)\n", __FUNCTION__, callbacks.isAppActive());
    return callbacks.isAppActive();
    
}

fileEntry *listFiles(short *fileCount, char **dynamicMemoryBuffer){
    // Returns a malloc'ed fileEntry array, and puts the file count into *fileCount.
    // Also returns a pointer to the memory that holds the file names, so that it can also
    // be freed afterward.

    printf("%s: \n", __PRETTY_FUNCTION__);
    return NULL;
}

void nextKeyOrMouseEvent(rogueEvent *returnEvent, boolean textInput, boolean colorsDance) {
    // Darth:
    //  Returns a _relevant_ event back to Brogue.
    //  It seems the other implems only loop to flush or serve as an event accumulator. When an actual Brogue recognized event occurts, the functions returns.
    //  There is some expectation that this will sortof loop efficiently until an input occurs.
    
    *returnEvent = callbacks.getBrogueEvent(textInput, colorsDance);
}

boolean pauseForMilliseconds(short milliseconds) {
    // Returns true if the player interrupted the wait with a keystroke or mouse action; otherwise false.
    
    //printf("%s(%i)\n",  __FUNCTION__, milliseconds);
    boolean res = callbacks.isEventWhilePaused(milliseconds);
    if (res) {
        printf("%i %s(%i)\n",res,  __FUNCTION__, milliseconds);
    }
    
    return (0 || res);
}

void plotChar(uchar inputChar,
              short xLoc, short yLoc,
              short foreRed, short foreGreen, short foreBlue,
              short backRed, short backGreen, short backBlue) {

//void plotChar(uchar inputChar,
//              short xLoc, short yLoc,
//              short backRed, short backGreen, short backBlue,
//              short foreRed, short foreGreen, short foreBlue) {
    //  plotChar: plots inputChar at (xLoc, yLoc) with specified background and foreground colors.
    //  Color components are given in ints from 0 to 100.
    
    PlotCharStruct charToPlot;

    charToPlot.inputChar = inputChar;
    charToPlot.xLoc = xLoc;
    charToPlot.yLoc = yLoc;
    charToPlot.backRed = backRed;
    charToPlot.backGreen = backGreen;
    charToPlot.backBlue = backBlue;
    charToPlot.foreRed = foreRed;
    charToPlot.foreGreen = foreGreen;
    charToPlot.foreBlue = foreBlue;
    
    callbacks.plotChar(charToPlot);
}

boolean saveHighScore(rogueHighScoresEntry theEntry) {
    // saves the high scores entry over the lowest-score entry if it qualifies.
    // returns whether the score qualified for the list.
    // This function ignores the date passed to it in theEntry and substitutes the current
    // date instead.

    printf("%s: \n", __PRETTY_FUNCTION__);
    return false;
}

#ifdef USE_CLIPBOARD
char *getClipboard() {
    // Returns a pointer to a char* containing the contents of the clipboard
    
    printf("%s: \n", __PRETTY_FUNCTION__);
    return 0;
}
#endif
